### PR TITLE
Strip LazyRef before calling simplified, in MT reduction

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -3248,7 +3248,20 @@ class TrackingTypeComparer(initctx: Context) extends TypeComparer(initctx) {
...
+            // by default will rewrap a LazyRef when applying its function.
+            // The result of those two things means that given a big enough input
+            // that recursive enough times through one or multiple match types,
+            // reducing and simplifying the result of the case bodies,
+            // can end up with a large stack of directly-nested lazy refs.
+            // And if that nesting level breaches `Config.LogPendingSubTypesThreshold`,
+            // then TypeComparer will eventually start returning `false` for `isSubType`.
+            // Or, under -Yno-deep-subtypes, start throwing AssertionErrors.
+            // So, we eagerly strip that lazy ref here to avoid the stacking.
+            tp.stripLazyRef.simplified
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3264]: Have you tried calling stripLazyRef in simplified?

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala]: Good idea.  It works.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 2]
```diff
@@ -3248,6 +3248,19 @@ class TrackingTypeComparer(initctx: Context) extends TypeComparer(initctx) {
...
+            // In addition to that, TypeMaps, such as the one that backs TypeOps.simplify,
+            // by default will rewrap a LazyRef when applying its function.
+            // The result of those two things means that given a big enough input
+            // that recursive enough times through one or multiple match types,
+            // reducing and simplifying the result of the case bodies,
+            // can end up with a large stack of directly-nested lazy refs.
+            // And if that nesting level breaches `Config.LogPendingSubTypesThreshold`,
+            // then TypeComparer will eventually start returning `false` for `isSubType`.
+            // Or, under -Yno-deep-subtypes, start throwing AssertionErrors.
+            // So, we eagerly strip that lazy ref here to avoid the stacking.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3263]: That comment is a bit out of place now, I would move it near the call to stripLazyRef

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 3]
```diff
@@ -1984,7 +1984,11 @@ object Types {
      *  this method handles this by never simplifying inside a `MethodicType`,
      *  except for replacing type parameters with associated type variables.
      */
-    def simplified(using Context): Type = TypeOps.simplify(this, null)
+    def simplified(using Context): Type =
+      // stripping LazyRef is important for the reduction of applied match types
+      // see the comment in matchCases/recur for more details
+      val tp = stripLazyRef
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala:1990]: I would move it in TypeOps.simplify

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/Types.scala:1990]: Recursively I think is a problem but I'll check

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala (hunk 2 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:3263]: Ok

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 3 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/Types.scala:1990]: Yeah, doesn't work:

```
27 suites passed, 3 failed, 30 total
    tests/neg/i15158.scala failed
    tests/neg-custom-args/isInstanceOf/i17435.scala failed
    tests/neg-custom-args/allow-deep-subtypes/i5877.scala failed
    tests/pos/i6146.scala failed
    tests/pos/i11064.scala failed
    tests/pos/i10900.scala failed
    tests/pos/i6146.scala failed
    tests/pos/i11064.scala failed
    tests/pos/i10900.scala failed
 ```

**Pull request merged**